cmd/gardener-admission-controller/app/gardener_admission_controller.go

```diff
@@ -0,0 +1,221 @@
+type options struct {
+	// ConfigFile is the location of the Gardener controller manager's configuration file.
+	ConfigFile string
+	scheme     *runtime.Scheme
+	codecs     serializer.CodecFactory
+}
+
+// addFlags adds flags for a specific Gardener controller manager to the specified FlagSet.
+func (o *options) addFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&o.ConfigFile, "config", o.ConfigFile, "Path to configuration file.")
+}
+
+// newOptions returns a new Options object.
+func newOptions() (*options, error) {
+	o := &options{
+		scheme: runtime.NewScheme(),
+	}
+
+	o.codecs = serializer.NewCodecFactory(o.scheme)
+
+	if err := config.AddToScheme(o.scheme); err != nil {
+		return nil, err
+	}
+	if err := admissioncontrollerconfigv1alpha1.AddToScheme(o.scheme); err != nil {
+		return nil, err
+	}
+
+	return o, nil
+}
+
+// loadConfigFromFile loads the contents of file and decodes it as a
+// ControllerManagerConfiguration object.
+func (o *options) loadConfigFromFile(file string) (*config.AdmissionControllerConfiguration, error) {
+	data, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+	return o.decodeConfig(data)
+}
+
+func (o *options) configFileSpecified() error {
+	if len(o.ConfigFile) == 0 {
+		return fmt.Errorf("missing Gardener controller manager config file")
+	}
+	return nil
+}
+
+// Validate validates all the required options.
+func (o *options) validate(args []string) error {
+	if len(args) != 0 {
+		return errors.New("arguments are not supported")
+	}
+
+	return nil
+}
+
+// decodeConfig decodes data as a ControllerManagerConfiguration object.
+func (o *options) decodeConfig(data []byte) (*config.AdmissionControllerConfiguration, error) {
+	configObj, gvk, err := o.codecs.UniversalDecoder().Decode(data, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	config, ok := configObj.(*config.AdmissionControllerConfiguration)
+	if !ok {
+		return nil, fmt.Errorf("got unexpected config type: %v", gvk)
+	}
+	return config, nil
+}
+
+func (o *options) applyDefaults(in *config.AdmissionControllerConfiguration) (*config.AdmissionControllerConfiguration, error) {
+	external, err := o.scheme.ConvertToVersion(in, admissioncontrollerconfigv1alpha1.SchemeGroupVersion)
+	if err != nil {
+		return nil, err
+	}
+	o.scheme.Default(external)
+
+	internal, err := o.scheme.ConvertToVersion(external, config.SchemeGroupVersion)
+	if err != nil {
+		return nil, err
+	}
+	out := internal.(*config.AdmissionControllerConfiguration)
+
+	return out, nil
+}
+
+// AdmissionController contains all necessary information to run the admission controller.
+type AdmissionController struct {
+	Config    *config.AdmissionControllerConfiguration
+	ClientMap clientmap.ClientMap
+}
+
+func (o *options) run(ctx context.Context) error {
+	c, err := o.loadConfigFromFile(o.ConfigFile)
+	if err != nil {
+		return err
+	}
+
+	if c, err = o.applyDefaults(c); err != nil {
```

Comment:
I am a bit confused by the version handling here...
Why are you first decoding the config in the internal version, then converting it back to the external one, default it and then convert it to the internal version again? Is this somehow required?

Reply:
This is the way you apply the defaults registered in the scheme for this kind. We did the same before in GCM, so I didn't question this handling.